rootston/layer_shell.c

```diff
@@ -0,0 +1,324 @@
+				layer->geo.x, layer->geo.y, 0);
+	}
+}
+
+static void unmap(struct wlr_layer_surface *layer_surface) {
+	struct roots_layer_surface *layer = layer_surface->data;
+	if (layer->link.prev) {
+		wl_list_remove(&layer->link);
+	}
+
+	struct wlr_output *wlr_output = layer_surface->output;
+	if (wlr_output != NULL) {
+		struct roots_output *output = wlr_output->data;
+		wlr_output_damage_add_box(output->damage, &layer->geo);
+	}
+}
+
+static void handle_destroy(struct wl_listener *listener, void *data) {
+	struct roots_layer_surface *layer = wl_container_of(
+			listener, layer, destroy);
+	if (layer->layer_surface->mapped) {
+		unmap(layer->layer_surface);
+	}
+	wl_list_remove(&layer->output_destroy.link);
+	wl_list_remove(&layer->output_mode.link);
+	arrange_layers(layer->layer_surface->output);
+	free(layer);
+}
+
+static void handle_map(struct wl_listener *listener, void *data) {
+	struct wlr_layer_surface *layer_surface = data;
+	struct roots_layer_surface *layer = layer_surface->data;
+	struct wlr_output *wlr_output = layer_surface->output;
+	struct roots_output *output = wlr_output->data;
+	wlr_output_damage_add_box(output->damage, &layer->geo);
+}
+
+static void handle_unmap(struct wl_listener *listener, void *data) {
+	struct roots_layer_surface *layer = wl_container_of(
+			listener, layer, unmap);
+	unmap(layer->layer_surface);
+}
+
+void handle_layer_shell_surface(struct wl_listener *listener, void *data) {
+	struct wlr_layer_surface *layer_surface = data;
+	struct roots_desktop *desktop =
+		wl_container_of(listener, desktop, layer_shell_surface);
+	wlr_log(L_DEBUG, "new layer surface: namespace %s layer %d anchor %d "
+			"size %dx%d margin %d,%d,%d,%d",
+		layer_surface->namespace, layer_surface->layer, layer_surface->layer,
+		layer_surface->client_pending.desired_width,
+		layer_surface->client_pending.desired_height,
+		layer_surface->client_pending.margin.top,
+		layer_surface->client_pending.margin.right,
+		layer_surface->client_pending.margin.bottom,
+		layer_surface->client_pending.margin.left);
+
+	struct roots_layer_surface *roots_surface =
+		calloc(1, sizeof(struct roots_layer_surface));
+	if (!roots_surface) {
+		return;
+	}
+
+	roots_surface->surface_commit.notify = handle_surface_commit;
+	wl_signal_add(&layer_surface->surface->events.commit,
+		&roots_surface->surface_commit);
+
+	roots_surface->output_destroy.notify = handle_output_destroy;
+	wl_signal_add(&layer_surface->output->events.destroy,
+		&roots_surface->output_destroy);
+
+	roots_surface->output_mode.notify = handle_output_mode;
+	wl_signal_add(&layer_surface->output->events.mode,
+		&roots_surface->output_mode);
+
+	roots_surface->destroy.notify = handle_destroy;
+	wl_signal_add(&layer_surface->events.destroy, &roots_surface->destroy);
+	roots_surface->map.notify = handle_map;
+	wl_signal_add(&layer_surface->events.map, &roots_surface->map);
+	roots_surface->unmap.notify = handle_unmap;
+	wl_signal_add(&layer_surface->events.unmap, &roots_surface->unmap);
```

Comment:
Need to listen for new subsurfaces, create `view_child` (might want to rename that), listen for commit and apply damage. See how other shells do this.